### PR TITLE
mkdir -p on file push

### LINKED ISSUE #2290: Uploading a file should (optionally) auto-create required directories
Using the command line client or the REST API, there should be an option or it should be standard behaviour that a file upload will create the missing directory structure. Alternatively, a new endpoint should be added to the API to allow creating a directory.

Example: using a fresh container
```
ubuntu@server:~$ lxc file push somefile test-container/tmp/nonexistingdirectory/somefile
error: open: No such file or directory
errno: 2
```

If this is not supported, offline file management is basically impossible as creating directories is a basic  required operation.

### PR DESCRIPTION
**Pull request by tych0**
Closes #2290

Signed-off-by: Tycho Andersen <tycho.andersen@canonical.com>

### CODE REVIEW

File: lxd/nsexec.go [hunk 1]
```diff
@@ -162,6 +162,18 @@ int manip_file_in_ns(char *rootfs, int pid, char *host, char *container, bool is
 	if (is_put && stat(container, &st) < 0)
 		exists = 0;
 
+	if (is_put) {
+		char *pos = strrchr(container, '/');
+		if (pos) {
+			*pos = 0;
+			if (mkdir_p(container, 0755) < 0) {
```
Comment (sean-jc) [inline lxd/nsexec.go]: Shouldn't this obey `mode` instead of using `0755`?

Comment (sean-jc) [inline lxd/nsexec.go]: I think this code would need to be moved inside the `!exists` check at [line 185](https://github.com/tych0/lxd/blob/fa2ecdf1c528f26809ba203cf4041c162ccba001/lxd/nsexec.go#L185) after configuring the uid/gid/mode defaults.

**Comment by stgraber** [comment]
So I don't think this is the right way to address this problem. As @sean-jc pointed out, we have a problem with the directory mode, specifically, we have no idea what it should be and no a typo in the path will cause the creation of additional directories that will be world readable.

I think we should stick to the "scp" behavior, that is, fail (with possibly a better error) if the path doesn't exist.

Then we should add the needed bits of API to allow for directory creation (and passing the needed mode and ownership info) which will make it possible to implement a "-r" flag to "lxc file push" (#1218).
This same API would also make it possible to create intermediary directories if the user really wants to, I'd however makes this an optional, non-default feature, lets call it "-p" (for consistency with mkdir).

File: lxd/nsexec.go (hunk 1 above)
Comment (tych0) [inline lxd/nsexec.go]: No, mode here is the mode of the actual file to be created. We need +x in the directory to be able to manipulate its children, so if the user specified mode is e.g. 600, subsequent mkdirs will fail. I suppose we could mask on +x to whatever the user specified mode is if we really wanted to.

**Comment by tych0** [comment]
Why not just have the client default to mkdir -p? We can have the api do it by not default, but I feel like the client should JFDI.

**Comment by stgraber** [comment]
None of the normal unix utilities do so (cp, scp, mv, ...) so it's not something users are very likely to expect. Most of the time where you get a "No such file or directory" error when using those, it's because you made a typo, not because you forgot to create a bunch of directories.

Having "lxc file push" silently create any missing directories for you will make it difficult for folks to figure out what just happened, thinking the file they wanted copied landed in the right place but then not finding it there because of their typo.

As I mentioned above, it can also be a potential security problem should someone for example type /rot/.ssh/id_rsa instead of /root/.ssh/id_rsa. If id_rsa is world readable on the source host, it will now be exposed to all users inside the container, rather than just root (as /root is 0700 or 0770 itself).

So I'm certainly not against having the option to do so, for folks who actually would like the one-command "mkdir -p" + "cp" behavior, but having it be the default feels like it would cause more problem than it would solve for 99% of our users.

**Comment by techtonik** [comment]
> None of the normal unix utilities do so (cp, scp, mv, ...) so it's not something users are very likely to expect. 

My 0.02 cents is that these normal unix utilities were not written in 21st century, and that those "safety measures" are outdated. I forgot when I used `rm` without `-rf` last time.

**Comment by srkunze** [comment]
@tych0 @techtonik I agree with you about the convenience part. It always sucks when creating/deleting on the shell. It feels so old. Most of the time, I don't care about directories but I need to hand-craft them all the time. -.- This also makes the charm of resource-based thinking like HTTP. Nobody cares about how many '/' (slashes) there are in the middle of a path.

Anyway.

@stgraber Isn't there a solution being both secure and convenient at the same time without yet another option?

At least to me it seems that LXD is resources-based like HTTP. So, it would be great if it could feel like it in almost all aspects. :)

**Comment by tych0** [comment]
Closing this in favor of #2351

**Pull request closed without merging**